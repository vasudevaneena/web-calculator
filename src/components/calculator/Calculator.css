.App{
  width: 300px;
  background: black;
  border-radius: 50px;
  color: white;
  margin: 50px; 
  padding: 20px;
  position: relative;
}
@media (min-width: 768px) {
  .App {
    
    width: 600px;
  }
}

.top {
  display: flex;
  /* height: 250px; */
  justify-content: space-between;
  padding: 0 20px;
}

.display {
  font-size: 75px;
  font-weight: 300;
  margin-bottom: 20px;
  text-align: right;
  word-wrap: break-word;
}

.buttons {
  display: grid;
  grid-gap: 20px;
  grid-template-columns: repeat(4, 1fr);
  grid-template-rows: repeat(5, 1fr);
}

.scientific-buttons {
  display: none;
}
@media (min-width: 768px) {
  .scientific-buttons {
    display: grid;
    grid-gap: 20px;
    grid-template-columns: repeat(4, 1fr);
    grid-template-rows: repeat(5, 1fr);
  }
}

.outer-buttons {
  display: grid;
  grid-gap: 20px;
  grid-template-columns: repeat(1, 1fr);
}
@media (min-width: 768px) {
  .outer-buttons {
    grid-template-columns: repeat(2, 1fr);
  }
}

.bottom {
  position: absolute;
  width: 200px;
  height: 5px;
  background: white;
  bottom: 10px;
  left: 50%;
  border-radius: 4px;
  transform: translateX(-50%);
}

button{
  border-color: black;
  background-color: black;
  color: white;
}
