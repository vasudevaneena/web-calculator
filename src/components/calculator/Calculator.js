import React, { useState } from "react";
import Button from "../button/Button";
import Scientificbutton from "../scientificButton/ScientificButton";
import commafy from "../utils/commafy";
import "./Calculator.css";

const Calculator = () => {
  const [value, setValue] = useState("0");
  const [memory, setMemory] = useState(null);
  const [operator, setOperator] = useState(null);
  let num;

  const handleButtonPress = (content) => () => {
    num = parseFloat(value);

    if (content === "AC") {
      setValue("0");
      setMemory(null);
      setOperator(null);
      return;
    }

    if (content === "±") {
      setValue((num * -1).toString());
      return;
    }

    if (content === "%") {
      setValue((num / 100).toString());
      setMemory(null);
      setOperator(null);
      return;
    }

    if (content === ".") {
      if (value.includes(".")) return;

      setValue(value + ".");
      return;
    }

    if (content === "+") {
      if (operator !== null) {
        setOperator(operator);
        if (operator === "+") {
          setMemory(memory + parseFloat(value));
        } else if (operator === "−") {
          setMemory(memory - parseFloat(value));
        } else if (operator === "×") {
          setMemory(memory * parseFloat(value));
        } else if (operator === "÷") {
          setMemory(memory / parseFloat(value));
        }
      } else {
        setMemory(parseFloat(value));
      }
      setValue("0");
      setOperator("+");
      return;
    }
    if (content === "−") {
      if (operator !== null) {
        if (operator === "+") {
          setMemory(memory + parseFloat(value));
        } else if (operator === "−") {
          setMemory(memory - parseFloat(value));
        } else if (operator === "×") {
          setMemory(memory * parseFloat(value));
        } else if (operator === "÷") {
          setMemory(memory / parseFloat(value));
        }
      } else {
        setMemory(parseFloat(value));
      }
      setValue("0");
      setOperator("−");
      return;
    }
    if (content === "×") {
      if (operator !== null) {
        if (operator === "+") {
          setMemory(memory + parseFloat(value));
        } else if (operator === "−") {
          setMemory(memory - parseFloat(value));
        } else if (operator === "×") {
          setMemory(memory * parseFloat(value));
        } else if (operator === "÷") {
          setMemory(memory / parseFloat(value));
        }
      } else {
        setMemory(parseFloat(value));
      }
      setValue("0");
      setOperator("×");
      return;
    }
    if (content === "÷") {
      if (operator !== null) {
        if (operator === "+") {
          setMemory(memory + parseFloat(value));
        } else if (operator === "−") {
          setMemory(memory - parseFloat(value));
        } else if (operator === "×") {
          setMemory(memory * parseFloat(value));
        } else if (operator === "÷") {
          setMemory(memory / parseFloat(value));
        }
      } else {
        setMemory(parseFloat(value));
      }
      setValue("0");
      setOperator("÷");
      return;
    }

    if (content === "=") {
      if (!operator) return;

      if (operator === "+") {
        setValue((memory + parseFloat(value)).toString());
      } else if (operator === "−") {
        setValue((memory - parseFloat(value)).toString());
      } else if (operator === "×") {
        setValue((memory * parseFloat(value)).toString());
      } else if (operator === "÷") {
        setValue((memory / parseFloat(value)).toString());
      }
      setMemory(null);
      setOperator(null);
      return;
    }

    if (value[value.length - 1] === ".") {
      setValue(value + content);
    } else {
      setValue(parseFloat(num + content).toString());
    }
  };

  return (
    <div className="App">
      <div className="top">
        <div className="memory">
         
        </div>
      </div>

      <div className="display" id="test-display">
        {memory}
        {operator}
        {commafy(value)}
      </div>

      <div className="outer-buttons">
        <div className="scientific-buttons">
          <Scientificbutton content="(" type="s_function" />
          <Scientificbutton content=")" type="s_function" />
          <Scientificbutton content="mc" type="s_function" />
          <Scientificbutton content="m+" type="s_function" />
          <Scientificbutton content="m-" type="s_function" />
          <Scientificbutton content="mr" type="s_function" />
          <Scientificbutton content="x!" type="s_function" />
          <Scientificbutton content="sin" type="s_function" />
          <Scientificbutton content="cos" type="s_function" />
          <Scientificbutton content="tan" type="s_function" />
          <Scientificbutton content="e" type="s_function" />
          <Scientificbutton content="EE" type="s_function" />
          <Scientificbutton content="Rad" type="s_function" />
          <Scientificbutton content="sinh" type="s_function" />
          <Scientificbutton content="cosh" type="s_function" />
          <Scientificbutton content="tanh" type="s_function" />
        </div>
        <div className="buttons">
          <Button
            onButtonClick={handleButtonPress}
            content="AC"
            type="function"
            id="test-ac"
          />
          <Button
            onButtonClick={handleButtonPress}
            content="±"
            type="function"
          />
          <Button
            onButtonClick={handleButtonPress}
            content="%"
            type="function"
            id="test-modulus"
          />
          <Button
            onButtonClick={handleButtonPress}
            content="÷"
            type="operator"
            id="test-division"
          />
          <Button
            onButtonClick={handleButtonPress}
            id="test-seven"
            content="7"
          />
          <Button
            onButtonClick={handleButtonPress}
            id="test-eight"
            content="8"
          />
          <Button
            onButtonClick={handleButtonPress}
            id="test-nine"
            content="9"
          />
          <Button
            onButtonClick={handleButtonPress}
            content="×"
            type="operator"
            id="test-multiply"
          />
          <Button
            onButtonClick={handleButtonPress}
            id="test-four"
            content="4"
          />
          <Button
            onButtonClick={handleButtonPress}
            id="test-five"
            content="5"
          />
          <Button onButtonClick={handleButtonPress} id="test-six" content="6" />
          <Button
            onButtonClick={handleButtonPress}
            content="−"
            type="operator"
            id="test-minus"
          />
          <Button onButtonClick={handleButtonPress} id="test-one" content="1" />
          <Button onButtonClick={handleButtonPress} id="test-two" content="2" />
          <Button
            onButtonClick={handleButtonPress}
            id="test-three"
            content="3"
          />
          <Button
            onButtonClick={handleButtonPress}
            content="+"
            type="operator"
            id="test-plus"
          />
          <Button onButtonClick={handleButtonPress} content="0" />
          <Button onButtonClick={handleButtonPress} content="." />
          <Button
            onButtonClick={handleButtonPress}
            content="="
            type="operator"
            id="test-equal"
          />
        </div>
      </div>

      <div className="bottom" />
    </div>
  );
};

export default Calculator;
