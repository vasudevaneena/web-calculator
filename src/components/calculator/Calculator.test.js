import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { configure,  mount } from "enzyme";
import Calculator from "./Calculator";

configure({ adapter: new Adapter() });
let wrapper;
beforeEach(() => {
  wrapper = mount(<Calculator />);
});
describe("component rendering", () => {
  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should render one <Button>", () => {
    expect(wrapper.find("Button")).toHaveLength(19);
  });
  it("should render one <ScientificButton>", () => {
    expect(wrapper.find("ScientificButton")).toHaveLength(16);
  });

  it("should render 1 <button>", () => {
    expect(wrapper.find("button")).toHaveLength(1);
  });
});

describe("Calculator display content test", () => {
  wrapper = mount(<Calculator />);
  wrapper.update();
  it("click button two -> should display 2 in display area", async () => {
    wrapper.find("#test-two").simulate("click");
    expect(wrapper.find("#test-display").props().children[2]).toEqual("2");
  });
  it("click operator + ", async () => {
    wrapper.find("#test-plus").simulate("click");
    expect(wrapper.find("#test-display").props().children[1]).toEqual("+");
  });

  it("click three", async () => {
    wrapper.find("#test-three").simulate("click");
    expect(wrapper.find("#test-display").props().children[2]).toEqual("3");
  });
});

describe("Calculator Addition functionalty test", () => {
  wrapper = mount(<Calculator />);
  wrapper.update();
  it("2 + 3 = 5", async () => {
    wrapper.find("#test-two").simulate("click");
    wrapper.find("#test-plus").simulate("click");
    wrapper.find("#test-three").simulate("click");
    wrapper.find("#test-equal").simulate("click");
    expect(wrapper.find("#test-display").props().children[2]).toEqual("5");
  });
});

describe("Calculator Minus functionalty test", () => {
  wrapper = mount(<Calculator />);
  wrapper.update();
  it("9 - 7 = 2", async () => {
    wrapper.find("#test-nine").simulate("click");
    wrapper.find("#test-minus").simulate("click");
    wrapper.find("#test-seven").simulate("click");
    wrapper.find("#test-equal").simulate("click");
    expect(wrapper.find("#test-display").props().children[2]).toEqual("2");
  });
});

describe("Calculator multiply functionalty test", () => {
  wrapper = mount(<Calculator />);
  wrapper.update();
  it("8 * 7 = 56", async () => {
    wrapper.find("#test-eight").simulate("click");
    wrapper.find("#test-multiply").simulate("click");
    wrapper.find("#test-seven").simulate("click");
    wrapper.find("#test-equal").simulate("click");
    expect(wrapper.find("#test-display").props().children[2]).toEqual("56");
  });
});

describe("Calculator division functionalty test", () => {
  wrapper = mount(<Calculator />);
  wrapper.update();
  it("9 / 3 = 3", async () => {
    wrapper.find("#test-nine").simulate("click");
    wrapper.find("#test-division").simulate("click");
    wrapper.find("#test-three").simulate("click");
    wrapper.find("#test-equal").simulate("click");
    expect(wrapper.find("#test-display").props().children[2]).toEqual("3");
  });
});

describe("Calculator modulus functionalty test", () => {
  wrapper = mount(<Calculator />);
  wrapper.update();
  it("7 % 3 = 0.073", async () => {
    wrapper.find("#test-seven").simulate("click");
    wrapper.find("#test-modulus").simulate("click");
    wrapper.find("#test-three").simulate("click");
    wrapper.find("#test-equal").simulate("click");
    expect(wrapper.find("#test-display").props().children[2]).toEqual("0.073");
  });
});

describe("clear the display", () => {
  wrapper = mount(<Calculator />);
  wrapper.update();
  it("reset", async () => {
    wrapper.find("#test-seven").simulate("click");
    wrapper.find("#test-modulus").simulate("click");
    wrapper.find("#test-three").simulate("click");
    wrapper.find("#test-ac").simulate("click");
    expect(wrapper.find("#test-display").props().children[2]).toEqual("0");
  });
});
