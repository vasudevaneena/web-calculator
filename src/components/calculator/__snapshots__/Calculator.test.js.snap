// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component rendering renders correctly 1`] = `ReactWrapper {}`;
